database/engine/pagecache.c

```diff
@@ -3,1311 +3,1062 @@
+
+                page = pd->page;
+                page_from_pd = true;
+                preloaded = pdc_page_status_check(pd, PDC_PAGE_PRELOADED);
+                waited = true;
+            }
+
+            if(!page || pdc_page_status_check(pd, PDC_PAGE_FAILED | PDC_PAGE_SKIP | PDC_PAGE_INVALID)) {
+                page = NULL;
+                continue;
+            }
         }
-    } else {
-        page_index = index;
-    }
-    pg_cache_reserve_pages(ctx, 1);
-
-    page_not_in_cache = 0;
-    uv_rwlock_rdlock(&page_index->lock);
-    int retry_count = 0;
-    while (1) {
-        descr = find_first_page_in_time_range(page_index, start_time_ut, end_time_ut);
-        if (NULL == descr || 0 == descr->page_length || retry_count == default_rrdeng_page_fetch_retries) {
-            /* non-empty page not found */
-            if (retry_count == default_rrdeng_page_fetch_retries)
-                error_report("Page cache timeout while waiting for page %p : returning FAIL", descr);
-            uv_rwlock_rdunlock(&page_index->lock);
-
-            pg_cache_release_pages(ctx, 1);
-            return NULL;
+
+        // we now have page
+
+        time_t page_start_time_s = pgc_page_start_time_t(page);
+        time_t page_end_time_s = pgc_page_end_time_t(page);
+        time_t page_update_every_s = pgc_page_update_every(page);
+        size_t page_length = pgc_page_data_size(main_cache, page);
+
+        if(unlikely(page_start_time_s == INVALID_TIME || page_end_time_s == INVALID_TIME)) {
+            __atomic_add_fetch(&rrdeng_cache_efficiency_stats.pages_zero_time_skipped, 1, __ATOMIC_RELAXED);
+            pgc_page_to_clean_evict_or_release(main_cache, page);
+            pdc_page_status_set(pd, PDC_PAGE_INVALID | PDC_PAGE_RELEASED);
+            pd->page = page = NULL;
+            continue;
         }
-        rrdeng_page_descr_mutex_lock(ctx, descr);
-        pg_cache_descr = descr->pg_cache_descr;
-        flags = pg_cache_descr->flags;
-        if ((flags & RRD_PAGE_POPULATED) && pg_cache_try_get_unsafe(descr, 0)) {
-            /* success */
-            rrdeng_page_descr_mutex_unlock(ctx, descr);
-            debug(D_RRDENGINE, "%s: Page was found in memory.", __func__);
-            break;
+        else if(page_length > RRDENG_BLOCK_SIZE) {
+            __atomic_add_fetch(&rrdeng_cache_efficiency_stats.pages_invalid_size_skipped, 1, __ATOMIC_RELAXED);
+            pgc_page_to_clean_evict_or_release(main_cache, page);
+            pdc_page_status_set(pd, PDC_PAGE_INVALID | PDC_PAGE_RELEASED);
+            pd->page = page = NULL;
+            continue;
         }
-        if (!(flags & RRD_PAGE_POPULATED) && pg_cache_try_get_unsafe(descr, 1)) {
-            struct rrdeng_cmd cmd;
+        else {
+            if (unlikely(page_update_every_s <= 0 || page_update_every_s > 86400)) {
+                __atomic_add_fetch(&rrdeng_cache_efficiency_stats.pages_invalid_update_every_fixed, 1, __ATOMIC_RELAXED);
```

Comment:
```sh
*** CID 433489:  High impact quality  (Y2K38_SAFETY)
/database/engine/pagecache.c: 869 in pg_cache_lookup_next()
863                 pd->page = page = NULL;
864                 continue;
865             }
866             else {
867                 if (unlikely(page_update_every_s <= 0 || page_update_every_s > 86400)) {
868                     __atomic_add_fetch(&rrdeng_cache_efficiency_stats.pages_invalid_update_every_fixed, 1, __ATOMIC_RELAXED);
>>>     CID 433489:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "page_update_every_s = pgc_page_fix_update_every(page, last_update_every_s)" is cast to "uint32_t".
869                     pd->update_every_s = page_update_every_s = pgc_page_fix_update_every(page, last_update_every_s);
870                 }
871     
872                 size_t entries_by_size = page_length / PAGE_POINT_CTX_SIZE_BYTES(ctx);
873                 size_t entries_by_time = (page_end_time_s - (page_start_time_s - page_update_every_s)) / page_update_every_s;
874                 if(unlikely(entries_by_size < entries_by_time)) {
```

Reply:
Ignore it. Update every cannot get that big.